### PR TITLE
Feature/staking strategy

### PR DESCRIPTION
**Pull request by mrtizmoatwork**

### CODE REVIEW

File: _i18n/en-US/_docs/effective_microengines.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+ACME Enterprises notices something suspicious on their network and wants to enrich their telemetry with intelligence produced by the PolySwarm marketplace.
+ACME submits an artifact to the PolySwarm marketplace via the PolySwarm API or a third party Ambassador.
+
+TODO: image
+
+ASIDE: when an Enterprise submits an artifact through the PolySwarm API, Swarm Technologies, Inc. acts as their Ambassador, handling the submission of the artifact and delivery of the results.
+
+The Ambassador acting on the Enterprise's behalf creates a bounty for the artifact and places a variable amount of NCT into the initial reward bucket for the artifact.
+For the purposes of illustration, let's assume 1 NCT was placed into the reward bucket.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: I would use a number larger than 1 NCT here. Something in the 2-6 range, but 1 is too small.

File: _i18n/en-US/_docs/effective_microengines.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+TODO: image
+
+ASIDE: the initial NCT placed on a bounty functions in a manner akin to Bitcoin or Ethereum transaction fees - it spurs initial economic interest in the artifact.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: I would remove the "akin to ... transaction fees". 1) we cannot assume that microengine dev's know how cryptocurrency transaction fees work. 2) we are adding confusion with our own marketplace transaction fees.

Just make a direct statement about what the bounty does in our marketplace.

File: _i18n/en-US/_docs/effective_microengines.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+Active PolySwarm engines are notified of this new bounty and the amount of NCT placed on the artifact.
+
+TODO: image
+
+Let's assume 8* engines (TODO: reference below ASIDE) find the initial NCT reward placed by the Ambassador to be sufficient for conducting a first-pass "triage" of the artifact.
+This lightweight "triage" step will determine whether the artifact falls in their area of expertise - more on this below.
+
+ASIDE: the number of active PolySwarm far beyond 8 and is growing by the day, but we'll keep this example simple for illustrative purposes.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: missing "microengines" after "active PolySwarm..."

File: _i18n/en-US/_docs/effective_microengines.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+Engines conduct their first-pass triage and determine:
+* Engines A, B, C, D, E: the artifact is within their area of expertise
+* Engines F, G, H: the artifact is outside their area of expertise
+
+Engines F, G and H ignore the bounty, choosing not to respond, whereas Engines A-E take a closer look.
+
+The engines conduct their full analysis of the artifact and derive the following results:
+* Engines A, C, D: malicious
+* Engines B, E: benign
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: These 3 lines should be removed. You are saying the same thing in the next section.

File: _i18n/en-US/_docs/effective_microengines.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+These engines telegraph their confidence in terms of the amount of NCT they stake, arriving at:
+* Engine A: 1 NCT / malicious
+* Engine B: 1 NCT / benign
+* Engine C: 2 NCT / malicious
+* Engine D: 1 NCT / malicious
+* Engine E: 2 NCT / benign
+
+Roughly speaking, Engines C and E are twice as confident in their assertion than the engines that agree with them.
+
+These assertions and their NCT stake amounts are streamed back to Ambassador at line speed.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: I think it is weird to say "streamed back.. at line speed" It's a computer network, of course it is.

How about we say "are immediately sent to the Ambassador." Simple, straight forward and clear.

File: _i18n/en-US/_docs/effective_microengines.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+The Ambassador optionally distills these assertions and delivers finished intelligence to ACME Enterprises for follow-on action.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: the distillation of assertions concept should be wording in simpler language.

something along the lines of this:
"The Ambassador analyzes these assertions and optionally combines them into a single piece of finished intelligence. That finished intelligence is then provided to ACME...."

File: _i18n/en-US/_docs/effective_microengines.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+ASIDE: The example presented here is a simplification of what would actually happen in the PolySwarm market.
+In the actual marketplace, far more engines would respond, stake amounts need not be integers and fees would be assessed by the marketplace in order to compensate Arbiters.
+
+
+## Monitor Your Engine's Marketplace Performance
+
+Any PolySwarm Portal user can track any engine's profit / loss from the Microengines page (TODO: link https://polyswarm.network/microengines/all).
+As an engine developer, you'll want to:
+1. keep an eye on competitors' performance (via the "All Microengines" tab) and
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: I think our recommendation should not be to keep track of competitors, but to keep track of their own performance when compared with competitors that can process the same file types.

File: _i18n/en-US/_docs/effective_microengines.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+
+### Close the Loop
+
+Visibility into underperforming engines allows the engine developer to take an underperforming engine offline as a means to stop loss.
+Still, without additional context, determining how to improve the engine before bringing it back online can be difficult.
+For this reason, *we strongly recommend that all engines archive the artifacts they have evaluated along with their assertion and NCT stake for any given artifact*.
+
+We're designing bookkeeping functions that make this a proactive process, helping the engine developer more quickly close the loop on engine improvement.
+As they become available, these features can be accessed via Portal on owned Microengines.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: We don't publish the word "Portal" as the name of the web application anywhere else. So I wouldn't use it here. Just call it "polyswarm.network".

File: _i18n/en-US/_docs/effective_microengines.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+This can be as simple as assigning a static weight per supported file type that influences the `confidence` score passed from the `Scanner` class.
+This type of strategy is better, but still is not ideal.
+3. *Confidence based on specific indicators*.
+This should be the goal of all well-performing engines.
+There are several engines on the PolySwarm marketplace that do this today, e.g. by dissecting Microsoft Word documents and identifying auto-executing Macro scripts that are known to be bad.
+This sort of artifact interrogation is optimal - it provides very high confidence signal that will allow these engine developers to develop an optimal staking strategy.
+
+TODO: code link example of file type based on confidence
+
+TODO: use IRIS-H as an example of #3?
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: I would not give away the inner workings of any specific engines unless those engine vendors have approved you doing this. Generalizing is ok.

File: _i18n/en-US/_docs/effective_microengines.md (hunk 1 above)
Comment (lm-ps) [reply _i18n/en-US/_docs/effective_microengines.md]: I don't think we need to talk about Ambassadors working on Enterprise's behalf, nor do we need to bring up that we often act as an Ambassador.

If y'all like it then let's keep it, though.

Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: I removed the reference to us being an Ambassador, but I think it's more illustrative / partial documentation to include the bits about Ambassador, so leaving that in.

Changed initial NCT amount to 5.

File: _i18n/en-US/_docs/effective_microengines.md (hunk 2 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: Removed the comparison to txn fees

File: _i18n/en-US/_docs/effective_microengines.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
+TODO: this page should fall under the Marketplace heading
+
+# Developing a Well-Performing Microengine
+
+As a Microengine developer, it's important to monitor your engine's performance and hone your NCT staking strategy to reflect your level of confidence concerning artifacts seen in the PolySwarm marketplace.
+Filtering which artifacts your engine asserts on and maintaining a strong correlation between amount staked and your confidence in your engine's assertion is essential for maintaining an effective (and profitable) Microengine.
```
[the remark this answers is not in the document]
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: changed all to microengine

File: _i18n/en-US/_docs/effective_microengines.md (hunk 4 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: removed

File: _i18n/en-US/_docs/effective_microengines.md (hunk 5 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: done

File: _i18n/en-US/_docs/effective_microengines.md (hunk 3 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: fixed

File: _i18n/en-US/_docs/effective_microengines.md (hunk 6 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: changed

File: _i18n/en-US/_docs/effective_microengines.md (hunk 8 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: removed references to portal

File: _i18n/en-US/_docs/effective_microengines.md (hunk 7 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: fixed

File: _i18n/en-US/_docs/effective_microengines.md (hunk 9 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: removed

File: _i18n/en-US/_docs/effective_microengines.md [hunk 11]
```diff
@@ -1,191 +1,248 @@
...
+  bid = int(min_bid + ((max_bid - min_bid) * avg_confidence))
+  
+  # Clamp bid between min_bid and max_bid
+  return max(min_bid, min(bid, max_bid))
+```
+TODO: end code aside
+
+CAPTION TODO: [The default `bid()` implementation](https://github.com/polyswarm/polyswarm-client/blob/5ea2d70a0df0de386a11cae45be40eaa9fedbd50/src/polyswarmclient/abstractmicroengine.py#L71) takes an array of `confidences` (one for each artifact in the bounty) into consideration when determining a stake amount.
+
+Microengine developers may choose to use this default staking strategy by deriving this `confidence` argument during the `scan()` function in their `Scanner` class - or can opt to override this logic entirely by subclassing the `TODO` function in their microengine.
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: what is a "TODO" function?

Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: my bad, meant bid

Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: fixed

File: _i18n/en-US/_docs/effective_microengines.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+  ...
+
+  async def scan(self, guid, content, chain):
+
+    # Reject files that libmagic does not identify as an ELF or PE by returning an empty ScanResult object
+    if not ((magic.from_buffer(content)[0:3] == "ELF") or (magic.from_buffer(content)[0:2] == "PE")):
+      return ScanResult()
+...
+```
+
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: @myhndl looks like this section has messed up code quote markers ^

File: _i18n/en-US/_docs/effective_microengines.md [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+    # Conduct the full analysis, arriving at a base confidence score
+    confidence_base = do_analysis()
+
+    ...
+
+    # Take file-based confidence into account when returning result
+    return ScanResult(bit=True, verdict=True, confidence=confidence_base+confidence_delta)
+...
+```
```
Comment (mrtizmoatwork) [inline _i18n/en-US/_docs/effective_microengines.md]: @myhndl this section also has messed up code quote markers. ^

File: _i18n/en-US/_docs/effective_microengines.md (hunk 12 above)
Comment (mrtizmoatwork) [reply _i18n/en-US/_docs/effective_microengines.md]: @joshcummingsdesign

Comment (mrtizmoatwork) [reply _i18n/en-US/_docs/effective_microengines.md]: @mrjoshc

File: _i18n/en-US/_docs/effective_microengines.md (hunk 13 above)
Comment (mrtizmoatwork) [reply _i18n/en-US/_docs/effective_microengines.md]: @mrjoshc

Comment (mrjoshc) [reply _i18n/en-US/_docs/effective_microengines.md]: @myhndl I think those are to imply that more code goes in between. Is that right? If so, I believe this is ready to go.

Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: yes, they're ellipses

File: _i18n/en-US/_docs/effective_microengines.md (hunk 12 above)
Comment (myhndl) [reply _i18n/en-US/_docs/effective_microengines.md]: ellipses

**Pull request merged**